service/storage_service.cc

```diff
@@ -5880,6 +5886,87 @@ future<> storage_service::add_tablet_replica(table_id table, dht::token token, l
+            if (tstate == topology::transition_state::tablet_draining ||
+                tstate == topology::transition_state::tablet_migration) {
+                break;
+            }
+            rtlogger.debug("del_tablet_replica(): topology state machine is busy: {}", tstate);
+            release_guard(std::move(guard));
+            co_await _topology_state_machine.event.wait();
+            guard = co_await _group0->client().start_operation(&_group0_as, raft_timeout{});
+        }
+
+        std::vector<canonical_mutation> updates;
+        auto& tmap = get_token_metadata().tablets().get_tablet_map(table);
+        auto tid = tmap.get_tablet_id(token);
+        auto& tinfo = tmap.get_tablet_info(tid);
+        auto last_token = tmap.get_last_token(tid);
+        auto gid = locator::global_tablet_id{table, tid};
+
+        if (tmap.get_tablet_transition_info(tid)) {
+            throw std::runtime_error(fmt::format("Tablet {} is in transition", gid));
+        }
+        auto* node = get_token_metadata().get_topology().find_node(dst.host);
+        if (!node) {
+            throw std::runtime_error(format("Unknown host: {}", dst.host));
+        }
+        if (dst.shard >= node->get_shard_count()) {
+            throw std::runtime_error(format("Host {} does not have shard {}", *node, dst.shard));
+        }
+
+        if (!locator::contains(tinfo.replicas, dst.host)) {
+            throw std::runtime_error(fmt::format("Tablet {} doesn't have replica on {}", gid, dst.host));
+        }
+
+        locator::tablet_replica_set new_replicas;
+        new_replicas.reserve(tinfo.replicas.size() - 1);
+        std::copy_if(tinfo.replicas.begin(), tinfo.replicas.end(), std::back_inserter(new_replicas), [&dst] (auto r) { return r != dst; });
+
+        updates.emplace_back(replica::tablet_mutation_builder(guard.write_timestamp(), table)
+            .set_new_replicas(last_token, new_replicas)
+            .set_stage(last_token, locator::tablet_transition_stage::allow_write_both_read_old)
+            .set_transition(last_token, locator::tablet_transition_kind::rebuild)
+            .build());
+        updates.emplace_back(topology_mutation_builder(guard.write_timestamp())
+            .set_transition_state(topology::transition_state::tablet_migration)
+            .set_version(_topology_state_machine._topology.version + 1)
+            .build());
+
+        sstring reason = format("Removing replica from tablet {}, node {}", gid, dst);
+        rtlogger.info("{}", reason);
+        rtlogger.trace("do update {} reason {}", updates, reason);
+        topology_change change{std::move(updates)};
+        group0_command g0_cmd = _group0->client().prepare_command(std::move(change), guard, reason);
+        try {
+            co_await _group0->client().add_entry(std::move(g0_cmd), std::move(guard), &_group0_as, raft_timeout{});
+            break;
+        } catch (group0_concurrent_modification&) {
```

Comment:
Should we add some short random delay before we retry to avoid multiple retires at the same time?

Reply:
There's a plan to add queue to add_entry(), but it's not short-term

Reply:
@kbr-scylla ?

Reply:
I am interested in how expensive it would be to drop and retry the prepare_command and add_entry in case of group0_concurrent_modification. If it is cheap, I guess it is fine to just do a simple random backoff and retry.  This kind of operation is rare anyway.

Reply:
It's going to be cheap indeed, but this way of retry is already in move_tablet() and in add_tablet_replica(). I'm going to generalize those in a followup, I can look at it then

Reply:
ok.

Reply:
> There's a plan to add queue to add_entry(), but it's not short-term

For group 0 there is no plan.

You may be thinking about topology coordinator request queue (https://github.com/scylladb/scylladb/issues/16822), which is different (higher-level abstraction) from add_entry. group0 add_entry allows only one concurrent operation, and there is no plan to change that.

Or you may be thinking about the "Raft for data path" ideas I presented on the summit, which will allow concurrent add_entry, but that's different than group 0.

---

Anyway, there is no need for a backoff mechanism for group0_concurrent_modifications. That's because when doing concurrent group 0 requests, at least one will always proceed without the exception. Also, these situations are very rare; e.g. when trying to execute a schema change while topology coordinator tries to do some transition.